[Feature] 메인화면 UI 구현

leeari95 opened the pull request:
### 📕 Issue

> 관련 이슈를 연결합니다.

x


### 📙 작업 내역

> 구현 내용 및 작업 했던 내역, 변경 사항을 포함합니다.

- [x] 날씨 정보를 나타내는 헤더뷰와 셀 구현
- [x] 코디와 스타일 목록을 나타내는 헤더뷰와 셀 구현
- [x] CollectionView 추가
	- [x] DiffableDataSource 적용
	- [x] CompositionalLayout 적용
- [x] 탭바 구현하기
- [x] 아이콘, 컬러 등을 추가


### 📘 작업 유형

- [x] 신규 기능 추가
- [ ] 버그 수정
- [ ] 리펙토링
- [ ] 문서 업데이트


### 📋 체크리스트

- [x] Merge 하는 브랜치가 올바른가?
- [x] PR과 관련없는 변경사항이 없는가?
- [x] 내 코드에 대한 자기 검토가 되었는가?
- [ ] 변경사항이 효과적이거나 동작이 작동한다는 것을 보증하는 테스트를 추가하였는가?
- [ ] 새로운 테스트와 기존의 테스트가 변경사항에 대해 만족하는가?


### 📝 리뷰 노트

> 구현 시에 고민이었던 점들 혹은 특정 부분에 대한 의도가 있다면 PR 리뷰를 돕기 위해 서술합니다. 리뷰어에게 특정 부분에 대한 집중 혹은 코멘트를 요청하는 경우에도 같이 작성해주면 좋습니다.

* CollectionView를 구현하면서 DiffableDataSource, CompositionalLayout 둘다 처음 적용해본거라... 최대한 익혀서 해봤는데, 혹시나 미흡한 점이 있다면 피드백 부탁드립니다!
* 썬이 공유해주신 글을 참고하여 Coordinator 수정 및 TabCoordinator를 추가했는데, 이 부분도 꼼꼼히 봐주시면 감사하겠습니다.
* 테스트해볼 수 있도록 MainCoordinator를 통해 앱 실행 시 첫번째로 나올 수 있도록 했는데, 제거가 필요하다면 말씀해주세요.

### 🏞️ 스크린샷

> 작업 내용과 관련된 화면을 추가합니다.

![image](https://user-images.githubusercontent.com/75905803/205944470-8c793de1-280d-409c-b6e1-c821316d977b.png)

leeari95 committed:
🚚 [style] Common에 샘플 이미지, 아이콘, 컬러 등을 추가

- 메인화면에 보여줄 샘플 이미지 추가
- 탭바에 추가할 날씨 아이콘 추가
- 3가지 컬러 추가 (FT, FTtext, FTSecondaryText)

leeari95 committed:
✨ [feat] UIKit 기본 기능 확장 추가

leeari95 committed:
✨ [feat] 네비게이션 좌측 상단에 위치할 LocationView를 구현

leeari95 committed:
✨ [feat] 날씨 정보를 나타내는 뷰와 날씨 아이콘을 나타내는 Cell 구현

leeari95 committed:
✨ [feat] FooterView와 HeaderView를 추가

leeari95 committed:
✨ [feat] 스타일을 나타내는 Cell, 코디사진을 나타내는 Cell 구현

leeari95 committed:
♻️ [refactor] 커스텀 뷰에 final 키워드 추가

leeari95 committed:
🚚 [style] 탭바 아이콘 추가

- 기존 sun 아이콘 이름을 icon-sun 으로 변경

leeari95 committed:
✨ [feat] 날씨 정보를 나타내는 HeaderView를 구현

leeari95 committed:
✨ [feat] MainCoordinator, TabCoordinator 구현

- Coordinator 프로토콜에 프로퍼티 추가 정의
- swiftlint.yml 파일 수정
- AuthCoordinatorInterface 수정

leeari95 committed:
✨ [feat] CodyCell 내부에 프로필 정보 추가 구현

leeari95 committed:
🚚 [style] HeaderView, WeatherCell 내부 뷰 속성 수정

leeari95 committed:
Merge branch 'develop' of https://github.com/YAPP-Github/21st-iOS-Team-2-iOS into feature/main-ios2-74

leeari95 commented on Fitfty/Projects/MainFeed/Sources/Views/StyleCell.swift:
```diff
@@ -0,0 +1,84 @@
+            verticalFittingPriority: UILayoutPriority
+        ) -> CGSize {
+            return CGSize(width: intrinsicContentSize.width, height: intrinsicContentSize.height)
+        }
```
잉... 다들 리뷰 끝내주시면 한꺼번에 수정할게요....

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/0cde5787e3d271b07930889de4cfbdc13da08e84

leeari95 committed:
🔧 [chore] 코드 앞 공백 맞지 않는 부분 개선

Daltonicc commented on Fitfty/Projects/Common/Resources/Images.xcassets/icons/icon-profile.imageset/Contents.json:
```diff
@@ -0,0 +1,21 @@
+    },
+    {
+      "idiom" : "universal",
+      "scale" : "3x"
```
[P3] 아리~ 에셋들 피그마에서 export하실 때 스케일 2x, 3x도 같이해서 넣어주세요 !

leeari95 replied:
이거 Sing Scale로 수정하는건 어떻게 생각하시나용~?
다른 스케일이 사용될까...? 라는 의문이 생겨서요.

Daltonicc replied:
1x에 대한 스케일만 넣게되면 해상도가 좋은 기기(최신 iOS기기?)에서는 픽셀이 낮아서 깨져보일 수도 있어요 ! 그래서 가급적이면 넣으시는게 좋을 것 같습니당 아이콘은 작아서 괜찮지만 큰 이미지일 경우는 깨져보일 수 있어요

leeari95 replied:
옹... 그런 이유라면 확장자 자체를 .svg로 바꾸는게 좋지 않을까요? 
그렇다면 2x, 3x 등록을 따로 할필요가 없을 것 같아서요!

Daltonicc replied:
옹 svg파일은 생각안해봤네요 svg파일로 하면 싱글 스케일로 대응이될 것 같아요 ! 그러면 svg파일로 수정 부탁드립니다😄

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/8079a5d4bcd4b4c9b0625f28c43da709468b3cea
넵! 기존 png 파일은 제거하고 svg 확장자 파일로 추가하여 수정하였습니다!

Daltonicc commented on Fitfty/Projects/Coordinator/Sources/AuthCoordinator.swift:
```diff
@@ -21,6 +24,7 @@ final class AuthCoordinator: Coordinator {
     func start() {
         let viewController = makeAuthViewController()
         navigationController.pushViewController(viewController, animated: true)
+        finishDelegate?.coordinatorDidFinish(childCoordinator: self)
```
[질문] 요거 메인플로우로 바로 보내려고 임시로 넣어두신 코드인거져??

leeari95 replied:
넵 맞습니다!!! 뷰 확인을 위해서 추가해둔건데, 제거가 필요하다고 생각되신다면 말씀해주세요!

Daltonicc commented on Fitfty/Projects/Coordinator/Sources/TabCoordinator.swift:
```diff
@@ -0,0 +1,141 @@
+
+enum TabBarPage {
+    case weather
+    case createCody
```
[질문] 코디를 탭바페이지로 두는게 맞을까라는 생각이드는데,, 만약 탭바페이지로 구성하게되면 코디를 등록완료했을 때의 플로우(탭의 루트뷰로 pop으로 돌아가게될지?)가 잘떠오르지 않아요 아리 생각이 궁금합니닷

leeari95 replied:
음... 제가 질문 이해를 잘 못하는 것 같아요. 🥲
TabBarPage라는 타입을 상단이 아니라 하단에 위치해야할 것 같다는 말씀이신건지?
아니면 TabBarPage라는 타입 자체에 대한 의문인걸까요?
일단 TabBarPage를 만들어둔 의미는 각 탭 마다 (날씨, 코디등록, 프로필) 플로우가 있을거니까...
해당 탭에 대한 플로우가 별도로 존재하겠다고 생각해서 저런식으로 나누어준거였거든요!

Daltonicc replied:
아 제가 궁금했던 건 코디등록이 탭으로 구성되는게 맞을까였어요 ! 코디등록이 탭이라면 코디등록 탭에서 코디 등록을 완료한뒤 어떤 동작을 주려하는지도 궁금했구여 ! 우리가 일반적으로 생각한다면 코디를 등록한뒤에 해당 페이지가 닫혀야할텐데 탭으로 구성되어 있으면 해당 동작이 좀 어색하지 않을까라고 생각해서 질문드렸어요 !

leeari95 replied:
오.. 그렇네요. 코디등록이라면 탭바페이지보다는 모달형태로 띄워야하는게 맞는 방법인 것 같습니다!
이 경우에는 add 버튼을 눌렀을 때, 단순히 뷰컨트롤러를 추가해주는 방식이 아닌 present 형식으로 뷰 컨트롤러를 띄우면 되지 않을까요??

Daltonicc replied:
네네 고렇게 가면 좋을 것 같아요 !

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/bf3fe133c028a0c814d363cfe49726ed53b6fce8
기존 NavigationVC를 띄워주는 방식을 modal 형태로 띄우는 방식으로 보완했습니다.

Daltonicc commented on Fitfty/Projects/MainFeed/Sources/ViewControllers/MainViewController.swift:
```diff
@@ -0,0 +1,216 @@
+        colletionView.register(WeatherCell.self)
+        colletionView.register(StyleCell.self)
+        colletionView.register(CodyCell.self)
+        colletionView.register(FooterView.self, forSupplementaryViewOfKind: UICollectionView.elementKindSectionFooter)
```
[질문] 혹시 FooterView만 Kind값으로 className이 아닌 elementKindSectionFooter값을 주신 이유가 있을까여 궁금쓰

leeari95 replied:
다른 뷰는 같은 헤더의 다른 타입의 헤더뷰라서 구분이 필요할 것 같아서 className으로 kind값을 할당해주었는데요.
FooterView의 경우 내용은 없고, 단순히 색상만 입혀져있는 뷰라서 이름 대신 elementKindSectionFooter값을 활용해주었거든요.
통일성 있게 className으로 해주는 것이 좋을까용?

Daltonicc replied:
사실 딱히 중요한건 아니긴해요😅 그래도 가급적이면 일관되면 좋을 것 같숩니닷 ! 만약 다른 푸터가 나중에 추가되면 고유값으로 또 수정해줘야할 수도 있으니 !

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/848f11ea28e4db4cac22c55524e6499907c87876
일관성있게 보완하였습니다~ ㅎㅎ

Daltonicc commented on Fitfty/Projects/MainFeed/Sources/ViewControllers/MainViewController.swift:
```diff
@@ -0,0 +1,216 @@
+        dataSource = UICollectionViewDiffableDataSource<Section, UUID>(
+            collectionView: collectionView,
+            cellProvider: { collectionView, indexPath, _ in
+                switch indexPath.section {
```
[P3] 아리~ indexPath.section 값으로 cell을 분기하는 것도 좋지만 요렇게되면 코드를 읽는 다른사람들은 0번섹션이 어떤 섹션인지? 찾아가야하는 소요가 생길 수도 있고 또 만약 새로운 섹션이 중간에 추가되는 등 개발적으로 변화가 있을때, 유연한 대응이 힘들 수 있을 것 같아요! Section타입을 만들어두셨으니 고걸 활용해보시는게 어떨까 생각합니다 !~~

leeari95 replied:
오! 감사합니다!! 조언주신대로 한번 개선해볼게요! 😁

leeari95 replied:
오! 감사합니다!! 조언주신대로 한번 개선해볼게요! 😁

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/4a4ec9acb24d1d7dced36a402972b4f963c3051a
Section 타입 활용해서 switch문 개선하였습니다!

Daltonicc commented on Fitfty/Projects/MainFeed/Sources/ViewControllers/MainViewController.swift:
```diff
@@ -0,0 +1,216 @@
+        collectionView.dataSource = dataSource
+    }
+    
+    private func setUpCollectionView() {
```
[P3] 함수의 네이밍과 맞지않는 동작을 하는 메서드 같아요 !  applySnapshot 요런 네이밍으로 바꿔보시는게 어떨까 생각합니다 !

leeari95 replied:
엇.. 정말 그렇네요. 네이밍 추천까지... 감사합니당!!! 😆🙏🏻

leeari95 replied:
오.. 그렇네용! 바로 수정해보겠습니다!!

Daltonicc commented on Fitfty/Projects/MainFeed/Sources/ViewControllers/MainViewController.swift:
```diff
@@ -0,0 +1,216 @@
+        return UICollectionViewCompositionalLayout { (sectionNumber, _) -> NSCollectionLayoutSection? in
+            
+            switch sectionNumber {
+            case 0: // 날씨 아이콘 섹션
```
[P3] 요기도 위 코멘트와 동일하게 만들어두신 Section 타입 활용해보시는게 어떨까 생각합니다 ! (그러면 주석제거도 가능할 것 같아요)

leeari95 replied:
넵 이 부분도 개선해보겠습니다! 🙏🏻

leeari95 replied:
넵 이 부분도 개선해보겠습니다! 🙏🏻

https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/4a4ec9acb24d1d7dced36a402972b4f963c3051a
Section 타입 활용해서 switch문 개선하였습니다!

Daltonicc commented on Fitfty/Projects/MainFeed/Sources/ViewControllers/MainViewController.swift:
```diff
@@ -0,0 +1,216 @@
+                        ),
+                        elementKind: UICollectionView.elementKindSectionFooter, alignment: .bottom)
+                ]
+                return section
```
[P4] 스위치문에 섹션 레이아웃 코드가 다 들어가있는데 함수로 분리해보시는게 어떨까 생각합니다~!
makeWeatherSectionLayout() 요런 네이밍의 함수로요 !

leeari95 replied:
넵 수정하겠습니다 !! 👍👍👍

leeari95 replied:
오! 조언 감사드립니다!!! 적극 반영해보도록 할게요!
덕분에 이 부분 좀 깔끔해질 것 같네요!!! 👍

https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/4a4ec9acb24d1d7dced36a402972b4f963c3051a
메소드 분리하여 개선하였습니다~

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/4a4ec9acb24d1d7dced36a402972b4f963c3051a

Daltonicc commented on Fitfty/Projects/MainFeed/Sources/Views/CodyCell.swift:
```diff
@@ -0,0 +1,118 @@
+    
+    override func prepareForReuse() {
+        super.prepareForReuse()
+//        reset()
```
[질문] 요건 왜 주석처리 되어있는걸까여?

leeari95 replied:
뷰 바인딩 처리시 필요한 메서드라서 일단 구현은 했는데, 주석처리 안해두니까
스크롤할때 이미지가 사라지는게 신경쓰여서 임시로 해두었는데, 주석을 빼두는게 좋을까요?

leeari95 replied:
뷰 바인딩 처리시 필요한 메서드라서 일단 구현은 했는데, 주석처리 안해두니까
스크롤할때 이미지가 사라지는게 신경쓰여서 임시로 해두었는데, 주석을 빼두는게 좋을까요?

Daltonicc replied:
넵넵 가급적이면 디벨롭에 들어갈때 주석처리된 코드는 안들어가는게 좋을 것 같아요 !

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/200e91a21341c9782dd77626ceebb007cb6af512
주석 제거하였습니다 ㅎㅎ

Daltonicc commented on Fitfty/Projects/MainFeed/Sources/Views/WeatherInfoHeaderView.swift:
```diff
@@ -0,0 +1,40 @@
+}
+
+extension WeatherInfoHeaderView {
+    
```
[P5] 요건 좀 사소한거긴 하지만,, extension 이후 첫 줄을 띄우시거나 안띄우시거나 통일되었으면 좋겟숩니다 일관적인 코드 스타일을 위해..!

leeari95 replied:
그렇네요! 에단은 붙여두셨던데... 저는 띄우는게 뭔가 확 트인 것 같고[?] 그렇더라구용?
어떻게 생각하시나요?

leeari95 replied:
그렇네요! 에단은 붙여두셨던데... 저는 띄우는게 뭔가 확 트인 것 같고[?] 그렇더라구용?
어떻게 생각하시나요?

Daltonicc replied:
요건 그냥 개인스타일이라 개인이 작성하는 방식만 통일되면 될 것 같아요 ! 아리가 작성한 코드중에 붙여쓴 곳도 있고 띄워진 곳도 있어서 코멘트 남긴거였습니닷 저는 붙여쓰는걸 선호하긴하는데 저땜에 스타일을 일부러 맞추실 필요는 전혀없습니다 !!!

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/200e91a21341c9782dd77626ceebb007cb6af512
일관성 있게 수정하였습니다!

Daltonicc reviewed (commented):
아리 고생하셨어요~~~! 메인피드 되게 이쁘네요👍🏿 리뷰확인해주시고 코멘트별로 만약 수정한거 있으시면 각 코멘트에 수정한 커밋해시 남겨주세요 !!

leeari95 committed:
♻️ [refactor] Section 내부 이니셜라이저를 추가하여 코드 구조 개선

- 기존에 indexPath로 switch문 구성되어있었던 부분을 Section 타입을 활용하여 개선함
- compositionalLayout 부분을 메소드 단위로 나누어서 개선

leeari95 committed:
🚚 [style] 오타 수정

leeari95 committed:
🚚 [style] 메소드명 수정 setUpCollectionView > applySnapshot

leeari95 committed:
♻️ [refactor] MainViewController 이니셜라이저 수정

- 코디네이터를 이니셜라이저를 통해 주입시키도록 개선

dudtjs1021ej commented on Fitfty/Projects/Common/Sources/Extension/UIFont+extension.swift:
```diff
@@ -0,0 +1,25 @@
+//
+//  UIFont+extension.swift
```
[질문] FitftyFont.swift를 사용하면 extension UIFont를 따로 추가하지 않아도 될 것 같은데 FiftyFont와 다르게 쓰이나요?😯

leeari95 replied:
음... 제가 커스텀 폰트를 사용한 경험이 없어서 그런 것 같아요.
일단 이 확장은 기본 다이나믹 타입 스타일에 bold체, 혹은 semibold 등등 다른 weight를 주고 싶을 때 사용할 수 있도록 만든 기능이거든요! 
이 경우는 에단이 만들어둔 커스텀 폰트를 사용하는 것이 맞는걸까요?

Daltonicc replied:
Fitfty폰트에 weight 대응을 할 수 있게끔 만들어드릴 수는 있는데 개인적으로 제 생각은 요런 UI적인 것들을 꼭 통일하기보다는(성능적으로 문제가 있지않는 한) 본인이 익숙한 스타일대로 취향껏 하시는걸 추천드립니당(저도 제가 익숙한대로 할거라서ㅎㅎ,,,)

leeari95 replied:
넵 그렇다면 따로 수정하진 않을게요 ㅎㅎ 😁

dudtjs1021ej commented on Fitfty/Projects/MainFeed/Sources/Views/HeaderView.swift:
```diff
@@ -0,0 +1,41 @@
+    private lazy var titleLabel: UILabel = {
+        let label = UILabel()
+        label.textColor = .label
+        label.font = .preferredFont(for: .title2, weight: .bold)
```
[p4] `label.font = FitftyFont.SFProDisplaySemibold(size: 24).font`
 폰트는 이렇게 통일하는건 어떨까요?

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6#discussion_r1045173297

위 코멘트 확인해주세요. 수정하지 않아도 괜찮을 것 같은데, 썬 생각은 어떠신가요?

dudtjs1021ej replied:
넵 좋아요! 각자 편한 스타일대로 하면 될듯합니다😙

dudtjs1021ej commented on Fitfty/Projects/Coordinator/Sources/TabCoordinator.swift:
```diff
@@ -0,0 +1,141 @@
+        }
+    }
+    
+    var pageIconImage: UIImage? {
+        switch self {
+        case .weather: return CommonAsset.Images.iconSun.image.withRenderingMode(.alwaysOriginal)
+        case .createCody: return CommonAsset.Images.iconAdd.image.withRenderingMode(.alwaysOriginal)
+        case .profile: return CommonAsset.Images.iconProfile.image.withRenderingMode(.alwaysOriginal)
+        }
+    }
+}
```
[p5] 지금 탭바는 프로필탭을 클릭해도 파랑색 아이콘으로 바뀌지 않는데 혹시 여유가 되신다면 이 부분 수정 부탁드립니다🥺

leeari95 replied:
아.. selected시 색상이 입혀졌었네요 ㅎㅎ 놓친 부분 알려주셔서 감사합니다!!! 🙏🏻

leeari95 replied:
https://github.com/YAPP-Github/21st-iOS-Team-2-iOS/pull/6/commits/2141ea729d2f6fcec49d03be89be550d4a3bbeb5
탭바 클릭 시 selected 효과가 날 수 있도록 개선하였습니다!

dudtjs1021ej reviewed (approved):
탭바랑 메인화면까지 고생하셨어요!! 아리 최고👍

leeari95 committed:
♻️ [refactor] icon 이미지 파일 추가 및 확장자 변경

- sun, profile 아이콘에 채워진 형태의 아이콘 추가

leeari95 committed:
✨ [feat] TabBar를 터치했을 때 아이콘이 변경되도록 추가

- TabBarPage 타입을 TabarController 타입 내부로 이동

leeari95 committed:
🚚 [style] 코드 컨벤션 일관성 있게 수정

leeari95 committed:
♻️ [refactor] FooterView의 kind를 수정

leeari95 committed:
🚚 [style] 코드 컨벤션 수정

leeari95 committed:
✨ [feat] 하단 탭바에 '+' 버튼 터치시 modal 형태로 화면을 띄우도록 추가

- 기존 TabBarController에 delegate 메소드 TabCoordinator로 이동
- TabBarPage 내부에 selectedIconImage 프로퍼티 추가

Daltonicc reviewed (approved):
고생하셧습니다 !
